% \VignetteEngine{knitr::knitr}
% \VignetteIndexEntry{1: R / Bioconductor Introduction}

\documentclass{beamer}
\usepackage{BioconductorSlides}
\title{\emph{useR!} 2015: Introduction to \R{} and \Bioconductor{}}
\author{Martin Morgan 
  (\href{mailto:mtmorgan@fredhutch.org}{mtmorgan@fredhutch.org}) \\
  Fred Hutchinson Cancer Research Center \\
  Seattle, WA, USA
}
\date{30 June, 2015}

\begin{document}

<<setup, echo=FALSE>>=
suppressPackageStartupMessages({
    library(ggplot2)
})
@ 

\maketitle

\begin{frame}{Outline}
  Content
  \begin{enumerate}
  \item Introductory presentation
  \item Follow-along labs
  \end{enumerate}
  Setup
  \begin{itemize}
  \item Amazon machine instance
  \item Github package:
    \href{http://github.org/Bioconductor/useR2015bc}{useR2015bc}
  \end{itemize}
\end{frame}

\section{R}

\begin{frame}[fragile]{\R: Statistical Computing Environment}

<<R>>=
1 + 2                       # calculator
x <- rnorm(1000)            # vectors, statistical
y <- x + rnorm(1000, sd=.8) # vectorized calculation
df <- data.frame(x=x, y=y)  # object construction
fit <- lm(y ~ x, df)        # linear model, formula
tbl <- anova(fit)           # class, generic, method
class(fit)                  # discovery
@ 

\end{frame}

\begin{frame}[fragile]{\R: Statistical Computing Environment}
  
  \begin{columns}
    \column{.5\textwidth}

<<R-plot>>=
plot(y ~ x, df, cex.lab=2)
abline(fit, col="red", 
       lwd=2)
@ 

    \column{.5\textwidth}

<<ggplot>>=
library(ggplot2)
ggplot(df, aes(x, y)) + 
    geom_point() + 
    stat_smooth(method="lm")
@ 
  \end{columns}
  
\end{frame}

\begin{frame}{\R: help!}
  \begin{itemize}
  \item \Rcode{?data.frame}
  \item $\Rightarrow$ \Rcode{methods(lm)}, \Rcode{methods(class=class(fit))}
  \item \Rcode{?"plot<tab>"}
  \item \Rcode{help(package="Biostrings")}
  \item $\Rightarrow$ \Rcode{vignette(package="GenomicRanges")}
  \end{itemize}
\end{frame}

\section{Bioconductor}

\begin{frame}{\Bioconductor{}}
  Analysis \& comprehension of high-throughput genomic data
  \begin{itemize}
  \item $>12$ years old; 1024 packages; widely used
  \item Sequencing (RNAseq, ChIPseq, variants, copy number, \ldots),
    microarrays, flow cytometery, proteomics, \ldots
  \item \url{http://bioconductor.org}, \url{https://support.bioconductor.org}
  \end{itemize}
  Themes
  \begin{itemize}
  \item Interoperable -- classes to work with genome-scale data,
    shared (where possible!) across packages
  \item Usable -- package vignettes, man pages, examples, \ldots
  \item Reproducible -- `release' and `devel' versions, updated every
    6 months
  \end{itemize}
\end{frame}

\begin{frame}{\Bioconductor: \Biocpkg{GenomicRanges}}
  \includegraphics[width=\textwidth, height=!]{our_figures/GRanges}
  \begin{itemize}
  \item Data: aligned reads, called peaks, SNP locations, CNVs, \ldots
  \item Annotation: gene models, variants, regulatory regions, \ldots
  \item \Rcode{findOverlaps()}, \Rcode{nearest()}, and many other
    useful range-based operations.
  \end{itemize}
\end{frame}

\begin{frame}[fragile]{\Bioconductor: \Biocpkg{SummarizedExperiment}}
  \begin{columns}
    \column{.5\textwidth}
    \includegraphics{our_figures/SE_Description.png}
    \column{.5\textwidth}
    Regions of interest $\times$ samples
    \begin{itemize}
    \item \Rcode{assay()} -- \Rcode{matrix}, e.g., counts of reads
      overlapping regions of interest.
    \item \Rcode{rowData()} -- regions of interest as \Rcode{GRanges}
      or \Rcode{GRangesList}
    \item \Rcode{colData()} -- \Rcode{DataFrame} describing samples.
    \end{itemize}
  \end{columns}
  \medskip\par
<<se, eval=FALSE>>=
se[,se$Treatment == "Control"]  # Control counts
@     
\end{frame}

\begin{frame}{\Bioconductor: Resources}
  \url{http://bioconductor.org}
  \begin{itemize}
  \item Packages -- \href{}{biocViews}, landing pages (e.g.,
    \Biocpkg{AnnotationHub})
  \item Course \& conference material; work flows; publications
  \item Developer resources
  \end{itemize}
  \url{https://support.bioconductor.org}
  \begin{itemize}
  \item Queston \& answer forum for users; usually fast, expert,
    friendly responses
  \item Contributed tutorials, news
  \end{itemize}
  Citations
  \begin{itemize}
  \item Huber et al. (2015) Orchestrating high-throughput genomic
    analysis with \Bioconductor. Nature Methods
    12:\href{http://dx.doi.org/10.1038/nmeth.3252}{115-121}.
  \item Lawrence et al. (2013) Software for Computing and Annotating
    Genomic Ranges. PLoS Comput Biol 9(8):
    \href{http://dx.doi.org/10.1371/journal.pcbi.1003118}{e1003118}.
  \end{itemize}
\end{frame}

\begin{frame}{Acknowledgments}
  \begin{itemize}
  \item Core (Seattle): \textbf{Sonali Arora}, Marc Carlson, Nate Hayden,
    Valerie Obenchain, Herv\'e Pag\`es, Paul Shannon, Dan Tenenbaum.
  \item The research reported in this presentation was supported by
    the National Cancer Institute and the National Human Genome
    Research Institute of the National Institutes of Health under
    Award numbers U24CA180996 and U41HG004059, and the National
    Science Foundation under Award number 1247813. The content is
    solely the responsibility of the authors and does not necessarily
    represent the official views of the National Institutes of Health
    or the National Science Foundation.
  \end{itemize}
  \medskip\par
  \href{http://bioconductor.org/BioC2015/}{BioC 2015} Annual
  Conference, Seattle, WA, 20-22 July.
\end{frame}

\end{document}
